src/wallet.js

```diff
@@ -345,65 +345,64 @@ const wallet = {
    * @inner
    */
   async getTxHistory(addresses, resolve, reject, connection = null, store = null) {
-    let hasMore = true;
-    let firstHash = null;
-    let addressesToSearch = addresses;
-    let history = null;
-
-    while (hasMore === true) {
-      let response;
-      // Using the GET request we end up getting a 414 (URI too large) if we customize the gap limit to a
-      // huge number. I am using the POST API because only new full node versions will have this API, so
-      // it's good to keep compatibility with older behaviour
-      if (addressesToSearch.length > MAX_ADDRESSES_GET) {
-        response = await walletApi.getAddressHistoryForAwaitPOST(addressesToSearch, firstHash);
-      } else {
+    // Split addresses array into chunks of at most MAX_ADDRESSES_GET size
+    // this is good when a use case customizes the GAP_LIMIT (e.g. 4000) then we don't
+    // request /address_history with 4000 addresses
+    const addressesChunks = _.chunk(addresses, MAX_ADDRESSES_GET);
+    const lastChunkIndex = addressesChunks.length - 1;
+
+    for (let i=0; i<=lastChunkIndex; i++) {
+      let hasMore = true;
+      let firstHash = null;
+      let addressesToSearch = addressesChunks[i];
+      while (hasMore === true) {
+        let response;
         response = await walletApi.getAddressHistoryForAwait(addressesToSearch, firstHash);
-      }
-      const result = response.data;
-      let ret = null;
-
-      if (result.success) {
-        hasMore = result.has_more;
-
-        if (hasMore) {
-          // If has more data we set the first_hash of the next search
-          // and update the addresses array with only the missing addresses
-          firstHash = result.first_hash;
-          const addrIndex = addressesToSearch.indexOf(result.first_address);
-          if (addrIndex === -1) {
-            throw Error("Invalid address returned from the server.");
+        const result = response.data;
+        let ret = null;
+
+        if (result.success) {
+          hasMore = result.has_more;
+
+          if (hasMore || i !== lastChunkIndex) {
+            // Update storage data with new history
+            ret = this.saveNewHistoryOnStorage(null, null, result.history, undefined, connection, store);
```

Comment:
I'm wondering if we could just "accumulate" the history from all calls and later just call once to update storage. But I guess we're changing the behavior, so better do it on a different PR.

Reply:
I guess this could work and would make the code simpler. But I agree with you that we should keep this PR only for the gap limit change. I will leave a comment in the code with this suggestion